### PR TITLE
Function to unpack a ragged array into a list of regular arrays

### PR DESCRIPTION
**Pull request by milancurcic**
* [x] Implementation
* [x] Tests
* [x] Example use in docstring

This function unpacks a ragged array into a list of regular arrays determined by rowsize. In support of #70 and other trajectory-bound-aware functions.

It's implemented using Philippe's EarthCube presentation approach (Xarray case 3).

Before proceeding to complete this PR, I'd like to get feedback on whether you like the direction and the API.

This function works with NumPy arrays and Xarray DataArrays, however passing DataArrays appears to be about 2 orders of magnitude slower. Here's a small benchmark with a ragged Xarray Dataset with 100 trajectories:

```
In [192]: %timeit res = unpack(ds.lon, ds.rowsize)
9.62 ms ± 9.27 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [193]: %timeit res = unpack(ds.lon.values, ds.rowsize.values)
163 µs ± 64.7 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Example:

```python
from clouddrift.adapters.gdp import to_raggedarray
from clouddrift.analysis import velocity_from_position
from clouddrift.dataformat import unpack

ds = to_raggedarray(n_random_id=100).to_xarray()

for lon, lat, time in list(zip(
    unpack(ds.lon.values, ds.rowsize),
    unpack(ds.lat.values, ds.rowsize),
    unpack(ds.time.values, ds.rowsize)
)):
    u, v = velocity_from_position(lon, lat, time)
    plt.scatter(lon, lat, c=np.sqrt(u**2 + v**2), s=10)
plt.colorbar()
```

Figure_1

### CODE REVIEW

**Comment by philippemiron** [comment]
Not sure if this is really needed to be honest. And for large arrays you have to duplicate all the memory to do this loop. I guess something like this should work, and there is probably a better way to use `map` and apply the `velocity_from_position` to segments of the arrays.

```
from clouddrift.adapters.gdp import to_raggedarray
from clouddrift.analysis import velocity_from_position
from clouddrift.dataformat import unpack

ds = to_raggedarray(n_random_id=100).to_xarray()

index_traj = np.insert(np.cumsum(rowsize), 0, 0)
for i in range(0, len(rowsize)):
    r = index_traj[i:i+2]
    u, v = velocity_from_position(ds.lon[r], ds.lat[r], ds.lat.time[r])
    plt.scatter(lon, lat, c=np.sqrt(u**2 + v**2), s=10)
plt.colorbar()
```

**Comment by milancurcic** [comment]
I agree with you that it's not strictly needed, but the purpose of it is to abstract away an implementation detail (`rowsize`) from the user.

**Comment by selipot** [comment]
The goal here is to emulate what is done in jLab by @jonathanlilly  with [`cellsplit.m`](https://github.com/jonathanlilly/jLab/blob/master/jCell/cellsplit.m), a function that allows one to split individual trajectories saved in a Matlab cell array into several sub-trajectories (which we will call **segments**) and hence a bigger cell array, based on some criteria such as a time gap. I am not sure what is the best way to do this is as above. 

As an example, starting from the GDP ragged array, remembering that individual trajectories have sometimes gaps of 12h or longer. How to create a new ragged array and/or a new rowsize variable that will index contiguous hourly segments instead of continuous ids?  

Instead of unpack, let's call this modifier/function `segment`; perhaps the syntax is:

```
newrowsize = segment(ds.time.values, ds.rowsize,condition)
```

**Comment by milancurcic** [comment]
I like `segment` as well. However, the purpose of this function (as in, in this PR) is not exactly what cellsplit does. It is merely to unpack individual trajectories from a 1-d ragged array representation, regardless of whether they have gaps or not.

I think we should pursue the cellsplit functionality as well, but as a separate function, because these are two separate features--1) segmenting by trajectory and 2) segmenting by contiguous time intervals.

**Comment by selipot** [comment]
Ok so what's next?

**Review by selipot — APPROVED** [review]

**Comment by selipot** [comment]
Maybe it should be called `ragged_to_list` or `unpack_ragged`

File: clouddrift/dataformat.py [hunk 1]
```diff
@@ -377,3 +377,23 @@ def to_parquet(self, filename: str):
         """
         ak.to_parquet(self.to_awkward(), filename)
         return
+
+
+def unpack(ragged_array: np.ndarray, rowsize: np.ndarray[int]) -> list[np.ndarray]:
+    """Unpack a ragged array into a list of regular arrays.
+
+    Unpacking ragged_array as an np.ndarray is about 2 orders of magnitude
```
Comment (selipot) [inline clouddrift/dataformat.py:385]: Unpacking a np.ndarray ragged_array is ...

File: clouddrift/dataformat.py [hunk 2]
```diff
@@ -377,3 +377,23 @@ def to_parquet(self, filename: str):
...
+    faster than unpacking it as an xr.DataArray, so unles you need a DataArray
```
Comment (selipot) [inline clouddrift/dataformat.py:386]: faster than unpacking a xr.DataArray ragged array, so unless you need a DataArray

**Comment by milancurcic** [comment]
Cool. I'll finish up this PR then (tests and example in docstring) and open an issue for segmenting by time gaps. In light of that, let sort out the naming for these two functions so they don't conflict and they're each clear.

> Maybe it should be called ragged_to_list or unpack_ragged

I actually had `unpack_ragged` first, but then removed it as I thought it was redundant as it's in a ragged array module (and we should rename it as such). `unpack_ragged` is certainly more clear and specific, so I like that. I like `ragged_to_list` slightly less (3 words instead of 2 kind of thing).

So, let's go with `unpack_ragged` here and reserve `segment` for segmenting by time gap?

**Review by selipot — CHANGES_REQUESTED** [review]

**Comment by selipot** [comment]
> So, let's go with `unpack_ragged` here and reserve `segment` for segmenting by time gap?

`unpack_ragged` sounds good. I would like to keep `segment` for segmenting by any arbitrary  criteria, not only time.

**Comment by milancurcic** [comment]
I completed the remaining items and addressed the requested changes, so I'll go ahead and merge. Thanks all.

**Pull request merged**